### PR TITLE
feat(ts-sdk): add vault-secrets derivation + VP bearer-token primitives

### PR DESCRIPTION
**Pull request by jrwbabylonlab**
SDK foundation for the VP bearer-token auth flow. Ships new primitives
(vault-secrets HKDF-Expand pipeline, VpTokenProvider, JsonRpcClient
bearer injection, Pre-PegIn authAnchorHash) but doesn't consume them
yet — existing flows are runtime-identical.

### CODE REVIEW

File: packages/babylon-ts-sdk/src/tbv/core/clients/vault-provider/auth/serverIdentity.ts [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+  const sig = stripHexPrefix(proof.signature).toLowerCase();
+  if (sig.length !== SCHNORR_SIG_HEX_LEN || !HEX_RE.test(sig)) {
+    throw new ServerIdentityError(
+      `signature must be 64-byte Schnorr hex; got ${sig.length} chars`,
+      "invalid_signature_encoding",
+    );
+  }
+
+  // TODO: add full BIP-322 sig verify.
+}
```
Comment (gbarkhatov) [inline packages/babylon-ts-sdk/src/tbv/core/clients/vault-provider/auth/serverIdentity.ts:220]: **BLOCKING (P0): BIP-322 signature verification is entirely absent**

`verifyServerIdentity()` validates structure (pinned pubkey match, expiry, format) but performs zero cryptographic verification of the BIP-322 Schnorr signature. The signature field is accepted as long as it is 64-byte hex.

**Impact:** A MITM who can intercept the `auth_createDepositorToken` response can substitute an arbitrary `ephemeral_pubkey` paired with any 128-char hex string. Since `server_pubkey` is publicly readable from the on-chain registry, the attacker can forge the entire `ServerIdentityResponse`. This defeats the ephemeral-key binding the proof is designed to enforce.

The code has a TODO comment acknowledging this, but shipping auth infrastructure without signature verification creates a false sense of security. If this ships and consumers start relying on `verifyServerIdentity()`, the missing check becomes a latent vulnerability.

**Recommendation:** Either:

**(a)** Implement BIP-322 verification before merge, or

**(b)** Rename the function to something like `validateServerIdentityFormat()` and add prominent JSDoc warnings that this is structural validation only, making it impossible to confuse with actual cryptographic verification. Add a tracking issue for the real implementation.

File: packages/babylon-ts-sdk/src/tbv/core/clients/vault-provider/json-rpc-client.ts [hunk 2]
```diff
@@ -139,6 +226,22 @@ export class JsonRpcClient {
     method: string,
     params: TParams,
     signal?: AbortSignal,
+  ): Promise<TResult> {
+    try {
+      return await this.callOnce<TParams, TResult>(method, params, signal);
+    } catch (error) {
+      if (this.tokenProvider && isAuthExpiredError(error)) {
+        this.tokenProvider.invalidate();
+        return await this.callOnce<TParams, TResult>(method, params, signal);
```
Comment (gbarkhatov) [inline packages/babylon-ts-sdk/src/tbv/core/clients/vault-provider/json-rpc-client.ts:245]: **BLOCKING (P0): Reactive auth-expired retry bypasses `retryableFor` guard**

The `call()` method's auth-expired catch block retries unconditionally:

```ts
catch (error) {
  if (this.tokenProvider && isAuthExpiredError(error)) {
    this.tokenProvider.invalidate();
    return await this.callOnce(method, params, signal); // retries ANY method
  }
  throw error;
}
```

This fires even for write methods explicitly excluded from `retryableFor`. A write method that returns `auth_expired` (e.g., due to a server bug or race condition) will be retried, potentially causing a double-submit.

**Recommendation:** Gate the reactive retry on `this.retryableFor(method)` or introduce a separate `authRetryableFor` predicate. At minimum, document the intentional design choice if double-submit is safe for all auth-gated methods.

File: packages/babylon-tbv-rust-wasm/src/index.ts [hunk 3]
```diff
@@ -48,7 +48,25 @@ export async function createPrePeginTransaction(
 ): Promise<PrePeginResult> {
   await initWasm();
 
-  const tx = new WasmPrePeginTx(
+  // The 13th positional arg `auth_anchor_hash` is an Option<String> in
+  // Rust — pass `undefined` for Pre-PegIns that do not commit an auth
+  // anchor. Requires a WASM build from a btc-vault commit ≥ 1ced81e5
+  // (btc-vault #1516).
+  const tx = new (WasmPrePeginTx as unknown as new (
```
Comment (gbarkhatov) [inline packages/babylon-tbv-rust-wasm/src/index.ts:55]: **IMPORTANT (P1): WASM version enforcement is build-discipline only**

The 13th `auth_anchor_hash` parameter to `WasmPrePeginTx` is passed via a type-unsafe cast (`WasmPrePeginTx as unknown as new (...)`). If a stale WASM binary is used, the parameter is silently ignored and no OP_RETURN is emitted.

The defensive check in `pegin.ts:188-210` catches this at runtime (good!), but this means the failure is at PSBT build time, not at import time. A CI check or WASM version assertion at init time would catch this earlier.

**Recommendation:** Consider adding a WASM version check in `initWasm()` that validates the WASM exports include the expected 13-arg constructor signature, or at minimum add a comment in CI config documenting the minimum WASM version requirement.

File: packages/babylon-ts-sdk/src/tbv/core/utils/fee/constants.ts [hunk 4]
```diff
@@ -44,23 +44,52 @@ export function rateBasedTxBufferFee(feeRate: number): number {
...
+ * @param vaultCount     - Number of vaults in the batch (≥1).
+ * @param hasAuthAnchor  - Whether the Pre-PegIn includes the auth-anchor
+ *                         OP_RETURN output. Defaults to `false`.
+ * @returns Total output count before change.
  */
-export function peginOutputCount(vaultCount: number): number {
-  return vaultCount + PEGIN_FIXED_OUTPUTS;
+export function peginOutputCount(
+  vaultCount: number,
+  hasAuthAnchor: boolean = false,
```
Comment (gbarkhatov) [inline packages/babylon-ts-sdk/src/tbv/core/utils/fee/constants.ts]: **IMPORTANT (P1): `peginOutputCount()` coupling requires caller discipline**

The `hasAuthAnchor` boolean defaults to `false`, which is correct for backward compatibility. However, the coupling between `buildPrePeginPsbt(authAnchorHash)` and `peginOutputCount(vaults, hasAuthAnchor)` is entirely caller-enforced. A mismatch silently underfunds the transaction.

The docstring warning is good. Consider whether `peginOutputCount` should take the `authAnchorHash` directly (truthy = has anchor) instead of a separate boolean, reducing the chance of caller error.

File: packages/babylon-ts-sdk/src/tbv/core/clients/vault-provider/auth/tokenProvider.ts [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+          now: this.now(),
+        });
+
+        const fresh: CachedToken = {
+          token: response.token,
+          expiresAt: response.expires_at,
+        };
+        this.cached = fresh;
+        return fresh;
+      } finally {
```
Comment (gbarkhatov) [inline packages/babylon-ts-sdk/src/tbv/core/clients/vault-provider/auth/tokenProvider.ts:210]: **IMPORTANT (P1): `inFlight` is correctly cleared on failure** ✅

The `finally` block ensures `this.inFlight = null` runs whether the acquire succeeds or fails, so a transient RPC failure does NOT permanently poison the provider. Good.

However, the test suite doesn't cover this scenario (failed acquire during single-flight). Consider adding a test where the RPC call rejects and verifying that a subsequent `getToken()` retries instead of returning the cached rejection.

File: packages/babylon-ts-sdk/src/tbv/core/vault-secrets/context.ts [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+  if (
+    !Number.isInteger(outpoint.vout) ||
+    outpoint.vout < 0 ||
+    outpoint.vout > 0xffffffff
+  ) {
+    throw new Error(`outpoint.vout must be a u32, got ${outpoint.vout}`);
+  }
+  const out = new Uint8Array(OUTPOINT_SIZE);
+  out.set(outpoint.txid, 0);
+  writeUint32BE(out, TXID_SIZE, outpoint.vout);
```
Comment (gbarkhatov) [inline packages/babylon-ts-sdk/src/tbv/core/vault-secrets/context.ts:73]: **IMPORTANT (P1): vout serialization endianness — verify cross-implementation compatibility**

The vout is serialized as big-endian (via `writeUint32BE`). This is intentional per the spec comment, but Bitcoin Core uses little-endian for outpoint serialization in raw transactions. Cross-implementation compatibility should be verified against the Rust `babe::vault_secrets` reference implementation to ensure byte-for-byte equivalence.

File: packages/babylon-ts-sdk/src/tbv/core/clients/vault-provider/auth/serverIdentity.ts [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+  // Validate both sides of the comparison are well-formed integers
+  // BEFORE comparing — untrusted JSON-RPC input can supply
+  // undefined/NaN/string values for `expires_at`, and relational
+  // comparisons with those silently evaluate to `false` (accepting the
+  // proof). Caller's `now` is injected but we still sanity-check it.
+  if (!Number.isSafeInteger(proof.expires_at)) {
+    throw new ServerIdentityError(
+      `expires_at must be a finite integer; got ${JSON.stringify(proof.expires_at)}`,
+      "expired",
```
Comment (gbarkhatov) [inline packages/babylon-ts-sdk/src/tbv/core/clients/vault-provider/auth/serverIdentity.ts]: **Suggestion (P2):** When `expires_at` is `NaN` or a string, the error reason is `"expired"`. A more precise reason like `"invalid_expires_at"` would help callers distinguish "proof was valid but expired" from "proof had garbage data". Same for the `now` check at line 119. Minor since this is internal SDK code.

File: packages/babylon-ts-sdk/src/tbv/core/clients/vault-provider/__tests__/json-rpc-client.test.ts [hunk 8]
```diff
@@ -324,4 +341,267 @@ describe("JsonRpcClient", () => {
     expect(body1.id).toBe(1);
```
Comment (gbarkhatov) [inline packages/babylon-ts-sdk/src/tbv/core/clients/vault-provider/__tests__/json-rpc-client.test.ts:341]: **Suggestion (P2): Test gaps**

1. **Concurrent auth-expired from multiple in-flight requests:** If two concurrent `call()` invocations both receive `auth_expired`, both will call `invalidate()` and retry. The single-flight in `VpTokenProvider` should deduplicate the re-acquire, but this race condition isn't tested.

2. **`isAuthExpiredError` with malformed `data`:** The function handles it correctly (returns `false`), but no test confirms behavior when `error.data` is a string, number, or array.

File: packages/babylon-ts-sdk/src/tbv/core/utils/fee/__tests__/constants.test.ts [hunk 9]
```diff
@@ -0,0 +1,38 @@
+import { describe, expect, it } from "vitest";
```
Comment (gbarkhatov) [inline packages/babylon-ts-sdk/src/tbv/core/utils/fee/__tests__/constants.test.ts:1]: **Suggestion (P2): Test gap — `peginOutputCount` edge cases**

`constants.test.ts` doesn't test edge cases like `vaultCount = 0` or negative values. These should either throw or be documented as valid.

File: packages/babylon-ts-sdk/src/tbv/core/clients/vault-provider/json-rpc-client.ts [hunk 10]
```diff
@@ -185,6 +297,13 @@ export class JsonRpcClient {
 
   /**
    * Make a JSON-RPC request returning the raw Response (unparsed body).
+   *
+   * Bearer tokens are injected identically to `call`. **Reactive refresh
+   * is NOT performed here** — the response body may be unbounded (e.g.
+   * claimer-artifact downloads), so the client refuses to parse it to
```
Comment (gbarkhatov) [inline packages/babylon-ts-sdk/src/tbv/core/clients/vault-provider/json-rpc-client.ts:313]: **Suggestion (P2):** `callRaw()` cannot trigger reactive refresh by design — well-documented here. Worth noting that callers downloading artifacts (claimer-artifact) via `callRaw()` must handle token refresh manually.

Also: on auth-expired retry (in `call()`), the `auth_anchor` preimage is transmitted twice (once in the original token request, once in the retry). This is bounded (one retry max) and mitigated by TLS, but worth noting in security documentation.

File: packages/babylon-ts-sdk/docs/api/clients.md [hunk 11]
```diff
@@ -694,9 +694,146 @@ Get the current pegout status from the vault provider daemon.
 
 ***
```
Comment (gbarkhatov) [inline packages/babylon-ts-sdk/docs/api/clients.md:696]: **Suggestion (P2):** These doc files appear auto-generated from JSDoc — verify they're not manually maintained to avoid merge conflicts with other in-flight PRs.

**Review by gbarkhatov — DISMISSED** [review]
**Verdict: Approve with comments**

Well-structured foundational PR. Code quality is high, test coverage is thorough, error handling follows project conventions, and security-critical paths have appropriate defensive assertions. The incomplete BIP-322 signature verification is the main concern but acceptable since no existing flow consumes these primitives yet.

See inline comments for details — 2 blocking items, 4 important items, 5 suggestions.

File: packages/babylon-tbv-rust-wasm/src/index.ts (hunk 3 above)
Comment (jrwbabylonlab) [reply packages/babylon-tbv-rust-wasm/src/index.ts:55]: The runtime check at pegin.ts:188-210 already catches the real failure mode by inspecting emitted script bytes. Tracking the CI .d.ts-snapshot check as a follow-up.

**Review by jeremy-babylonlabs — APPROVED** [review]
lgtm
